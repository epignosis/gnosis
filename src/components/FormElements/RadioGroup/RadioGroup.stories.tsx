import React from "react";
import { Story } from "@storybook/react";
import RadioGroupComponent, { RadioGroupProps } from "./RadioGroup";

export default {
  title: "components/Form Elements/Radio Group",
  component: RadioGroupComponent,
  args: {
    size: "md",
    inline: false,
    initialValue: "",
    groupname: "radio-group",
  },
  argTypes: {
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
    initialValue: {
      control: {
        type: "select",
        options: ["all", "progress", "completed", "failed"],
      },
    },
  },
};

export const RadioGroup: Story<RadioGroupProps> = (args) => <RadioGroupComponent {...args} />;

RadioGroup.args = {
  options: [
    {
      label: "All",
      value: "all",
    },
    {
      label: "In progress",
      value: "progress",
      disabled: true,
    },
    {
      label: "Completed",
      value: "completed",
    },
    {
      label: "Failed",
      value: "failed",
    },
  ],
};
