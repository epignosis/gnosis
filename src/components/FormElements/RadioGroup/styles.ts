import { css, SerializedStyles, Theme } from "@emotion/react";
import { InputSize } from "../Input/Input";

export const container = (inline: boolean): SerializedStyles => css`
  display: flex;
  flex-direction: ${inline ? "row" : "column"};
  padding: 0;
  border: 0;
  margin: 0;
`;

export const radioButtonContainer = (
  { typeScaleSizes, formElements: { checkbox } }: Theme,
  { size, inline }: { size: InputSize; inline: boolean },
): SerializedStyles => {
  const fontSizes = {
    md: typeScaleSizes.sm,
    lg: typeScaleSizes.md,
  };

  return css`
    padding-inline: 0 ${inline ? "1.5rem" : "0"};
    padding-block: 0.75rem;

    input[type="radio"] {
      opacity: 0;

      &:focus,
      &:hover {
        + label::before {
          box-shadow: 0px 0px 0px 9px ${checkbox.input.shadowColor};
        }
      }

      &:checked {
        + label::after {
          background-color: ${checkbox.input.background};
        }
      }

      &:disabled {
        + label {
          cursor: not-allowed;

          &::before {
            box-shadow: none;
          }

          &::after {
            background-color: ${checkbox.input.borderColor};
          }
        }
      }

      + label {
        font-size: ${fontSizes[size]};
        position: relative;
        display: inline-block;
        padding-inline-start: 1.5rem;
        cursor: pointer;

        &::before {
          content: "";
          position: absolute;
          inset-inline-start: -12px;
          display: inline-block;
          width: 20px;
          height: 20px;
          background-color: transparent;
          border: 1px solid ${checkbox.input.borderColor};
          border-radius: 50%;
          margin-inline-end: 1rem;
          transition: box-shadow 0.1s ease-out;
          top: 50%;
          transform: translateY(-50%);
        }

        &::after {
          content: "";
          position: absolute;
          display: inline-block;
          inset-inline-start: -9px;
          top: 50%;
          width: 14px;
          height: 14px;
          border-radius: 50%;
          transition: background-color 0.1s ease-out;
          transform: translateY(-50%);
        }
      }
    }
  `;
};
