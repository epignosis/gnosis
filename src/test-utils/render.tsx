import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
