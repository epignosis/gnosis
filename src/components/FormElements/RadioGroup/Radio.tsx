import React, { FC, InputHTMLAttributes } from "react";
import { SerializedStyles } from "@emotion/react";
import { InputSize } from "../Input/Input";
import { radioButtonContainer } from "./styles";
import { ExtendableProps } from "types/common";

export type RadioOption = {
  value: string;
  label: string;
  disabled?: boolean;
};

export type RadioProps = ExtendableProps<
  InputHTMLAttributes<HTMLInputElement>,
  RadioOption & {
    id: string;
    size?: InputSize;
    inline?: boolean;
  }
>;

const Radio: FC<RadioProps> = (props) => {
  const { id, label, size = "md", inline = false, ...rest } = props;

  return (
    <div css={(theme): SerializedStyles => radioButtonContainer(theme, { size, inline })}>
      <input id={id} type="radio" {...rest} />
      <label htmlFor={id}>{label}</label>
    </div>
  );
};

export default Radio;
