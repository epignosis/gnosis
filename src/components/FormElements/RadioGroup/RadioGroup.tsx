import React, { FC, ChangeEvent, useState, useEffect } from "react";
import Radio, { RadioOption } from "./Radio";
import { container } from "./styles";
import { InputSize } from "@components/FormElements/Input/Input";

export type RadioGroupProps = {
  id: string;
  groupname: string;
  options: RadioOption[];
  onChange?: (selectedValue: string) => void;
  initialValue?: string;
  inline?: boolean;
  size?: InputSize;
  className?: string;
};

const RadioGroup: FC<RadioGroupProps> = ({
  groupname,
  options,
  onChange,
  initialValue = "",
  inline = false,
  size = "md",
  id,
  className,
}) => {
  const [selectedValue, setSelectedValue] = useState(initialValue);
  const selectValue = (e: ChangeEvent<HTMLInputElement>): void => {
    const selectedValue = e.target.value;

    setSelectedValue(selectedValue);
  };

  useEffect(() => {
    onChange && onChange(selectedValue);
  }, [selectedValue, onChange]);

  return (
    <fieldset css={container(inline)} id={id || groupname} className={className}>
      {options.map((option) => (
        <Radio
          key={option.value}
          id={`${groupname}-${option.value}`}
          size={size}
          checked={selectedValue === option.value}
          name={groupname}
          onChange={selectValue}
          inline
          {...option}
        />
      ))}
    </fieldset>
  );
};

export default RadioGroup;
