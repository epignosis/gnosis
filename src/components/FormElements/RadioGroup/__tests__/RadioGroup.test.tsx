import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import RadioGroup from "../RadioGroup";
import { screen, render } from "@test-utils/render";

const OPTIONS = [
  {
    label: faker.company.companyName(),
    value: faker.company.companySuffix(),
  },
  {
    label: faker.commerce.department(),
    value: faker.commerce.productAdjective(),
  },
  {
    label: faker.git.commitMessage(),
    value: faker.git.commitSha(),
  },
  {
    label: faker.hacker.phrase(),
    value: faker.hacker.abbreviation(),
  },
];

describe("<RadioGroup />", () => {
  it("renders properly", () => {
    render(<RadioGroup groupname={faker.random.words()} options={OPTIONS} />);

    const radios = screen.getAllByRole("radio");

    expect(radios).toHaveLength(4);
  });

  it("renders with initial value selected", () => {
    render(
      <RadioGroup
        groupname={faker.random.words()}
        options={OPTIONS}
        initialValue={OPTIONS[2].value}
      />,
    );

    const thirdOption = screen.getByLabelText(OPTIONS[2].label);

    expect(thirdOption).toBeChecked();
  });

  it("changes selected value", () => {
    const mockFn = jest.fn();

    render(<RadioGroup groupname={faker.random.words()} options={OPTIONS} onChange={mockFn} />);

    const radios = screen.getAllByRole("radio");

    userEvent.click(radios[0]);

    expect(radios[0]).toBeChecked();

    userEvent.click(radios[2]);

    expect(radios[0]).not.toBeChecked();
    expect(radios[2]).toBeChecked();
    expect(mockFn).toHaveBeenCalledTimes(3);
  });

  it("doesn't select a disabled option", () => {
    render(<RadioGroup groupname={faker.random.words()} options={OPTIONS} />);

    const firstOption = screen.getByText(OPTIONS[1].label);
    const firstOptionLabel = screen.getByLabelText(OPTIONS[1].label);

    expect(firstOption).not.toBeChecked();

    userEvent.click(firstOptionLabel);

    expect(firstOption).not.toBeChecked();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <RadioGroup
        groupname="Test groupname"
        options={[{ label: "Test option label", value: "testOptionValue" }]}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with `inline = true`", () => {
    const { container } = render(
      <RadioGroup
        groupname="Test groupname"
        options={[{ label: "Test option label", value: "testOptionValue" }]}
        inline
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
