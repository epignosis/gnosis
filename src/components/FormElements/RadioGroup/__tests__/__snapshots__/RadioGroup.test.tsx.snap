// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> matches snapshot 1`] = `
<div>
  <fieldset
    class="css-amgjt9-container-container"
    id="testRadioGroup"
  >
    <div
      class="css-ud70d6-radioButtonContainer-radioButtonContainer"
    >
      <input
        id="Test groupname-testOptionValue"
        name="Test groupname"
        type="radio"
        value="testOptionValue"
      />
      <label
        for="Test groupname-testOptionValue"
      >
        Test option label
      </label>
    </div>
  </fieldset>
</div>
`;

exports[`<RadioGroup /> matches snapshot with \`inline = true\` 1`] = `
<div>
  <fieldset
    class="css-jgs0pf-container-container"
    id="testRadioGroup"
  >
    <div
      class="css-ud70d6-radioButtonContainer-radioButtonContainer"
    >
      <input
        id="Test groupname-testOptionValue"
        name="Test groupname"
        type="radio"
        value="testOptionValue"
      />
      <label
        for="Test groupname-testOptionValue"
      >
        Test option label
      </label>
    </div>
  </fieldset>
</div>
`;
